### DISCUSSION TITLE
Prisma unable to update db when enum is changed.

### DISCUSSION THREAD

**Discussion by joelrb** [prisma/prisma#23236] [Q&A] [answered]
### Bug description

Updated Enum from:

```
enum CompanyTransRole {
  Client 
}
```

to
```
enum CompanyTransRole {
  Customer 
}
```

and added new model 

```
model CompanyTransactionRole { 
  id                             Int                   @id @default(autoincrement())
  companyId             String
  transactionRole       CompanyTransRole        @default(Customer)
}
```

```
Error: P1014
The underlying table for model `sc_companies.CompanyTransactionRole` does not exist.
```

For this to work, I have to delete the Type in pgAdmin.

### Expected behavior

Database updated on db push.

### Prisma information

```prisma
// Add your schema.prisma
```
See bug description (self explanatory).
```ts
// Add your code using Prisma Client
```

### Environment & setup

- OS: Windows
- Database: PostgreSQL
- Node.js version: v21.6.2

### Prisma Version

```
5.9.1
```

**Comment by janpio** [comment]
What command exactly is returning this error?
What do you mean by "Type in pgAdmin"?

**Reply by joelrb** [reply]
What command exactly is returning this error?

On npx prisma db push. This error occurs.
Error: P1014
The underlying table for model `sc_companies.CompanyTransactionRole` does not exist.

What do you mean by "Type in pgAdmin"?
Found in the db tool Postgres Admin (https://www.pgadmin.org/)

The Types in the Schema (see attached)
![Screenshot 2024-03-04 232008](https://github.com/prisma/prisma/assets/225284/e5417ce7-43c0-4b23-9fd0-7dd99d6003d2)

The Type there is the Enum in the schema.prisma. It's also added to the Types on prisma db push.

**Reply by janpio** [reply]
Hm interesting, it seems to get confused in the SQL it generates and try to modify/remove something that does not exist yet.

Can you create a new project, and use `migrate dev` to record the steps of the migrations? So first create everything as when you started, then use `migrate dev` to record that. Then make your change, and use `migrate dev` again to record the SQL that it would run - that should be similar to `db push` and also fail the same way.

**Reply by joelrb** [reply]
The type CompanyTransRole already existed in the database before I added the model CompanyTransactionRole.

The issue is more on the enum change from Client to Customer. The `db push` error went away when I manually deleted the type from pgAdmin.

Ok I will create a new project and do `migrate dev`. Let you know!

**Comment by joelrb** [comment]
Here's result of new test. Could not replicate exactly the same way that cause the error in the issue I posted.

In this test, it appears migration or db push fails when there's data in the database.

**Schema**

```
enum UserRole {
  SUPERADMIN
  ADMIN
  USER
}

model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
  role                  UserRole               @default(USER)
}
```

1. **No data in database, migration succeeded:**

```
Applying migration `20240306043803_test1`

The following migration(s) have been created and applied from new schema changes:

migrations/
  └─ 20240306043803_test1/
    └─ migration.sql

Your database is now in sync with your schema.
```

2. **Changed enum, still no data in database:**

```
The following migration(s) have been created and applied from new schema changes:

migrations/
  └─ 20240306044005_test2/
    └─ migration.sql

Your database is now in sync with your schema.
```

3. **Added data in pgAdmin**, changed Enum in schema from USER to USERtest

npx prisma migrate dev

```
√ Enter a name for the new migration: ... test3
Applying migration `20240306044215_test3`
Error: ERROR: current transaction is aborted, commands ignored until end of transaction block
   0: schema_core::commands::apply_migrations::Applying migration
           with migration_name="20240306044215_test3"
             at schema-engine\core\src\commands\apply_migrations.rs:91
   1: schema_core::state::ApplyMigrations
             at schema-engine\core\src\state.rs:201
```

 npx prisma db push

```
√ Do you want to ignore the warning(s)? ... yes
Error: ERROR: invalid input value for enum "UserRole_new": "USER"
   0: sql_schema_connector::apply_migration::migration_step
           with step=AlterEnum(AlterEnum { id: MigrationPair { previous: EnumId(0), next: EnumId(0) }, created_variants: ["USERtest"], dropped_variants: ["USER"], previous_usages_as_default: [(TableColumnId(3), Some(TableColumnId(3)))] })
             at schema-engine\connectors\sql-schema-connector\src\apply_migration.rs:21
   1: sql_schema_connector::apply_migration::apply_migration
             at schema-engine\connectors\sql-schema-connector\src\apply_migration.rs:10
   2: schema_core::state::SchemaPush
             at schema-engine\core\src\state.rs:433
 ```

4. **Deleted data in pgAdmin**

migrate dev / db push succeeded.

**Reply by janpio** [reply]
What is the generated SQL of the migration that fails?
What is the state of the database schema and data (in SQL) at that moment?

**Reply by joelrb** [reply]
Hi, please see attached PDF which would make this clearer to you. I hope.

[DBtest.pdf](https://github.com/prisma/prisma/files/14510521/DBtest.pdf)

**Reply by joelrb** [reply]
Either we delete the data manually from DB or on migration do a reset db, which would be a disaster for production.

**Reply by janpio** [reply]
Ok, so you are changing an `enum` _value_.


Isn't that exactly what the warning you get warns about?
<img width="463" alt="image" src="https://github.com/prisma/prisma/assets/183673/85ebc671-b1c5-49c1-a632-08b9fb735622">
As long as the value `USERtest` is used in the data, you can not remove the definition without it failing.

Instead, add an _additional_ new value and put it as the default, but keep the old one around in the enum for this migration - that should allow you to run the migration. Then manually update the data to use the new value instead of the old (so the old one is not used any more) - and then do another migration where you remove the old, now obsolete and unused value. That should get rid of the warning, and then also get rid of the error.

**Reply by AhmadNadeemJB** [reply]
You guys found any solution for that? I'm currently facing the same issue i don't have any data in my table but it still shows me this error. 
![image_2024-08-26_060428861](https://github.com/user-attachments/assets/84accbd1-8c19-4b80-91c9-82b7ebdb7e6a)

I changed enum from 

```
enum RoomType {
  Single
  Double
  Suite
  Other
}
```
to
```
enum RoomType {
  StandardRoom
  DeluxeRoom
  Suite
  Other
}
```

**Reply by joelrb** [reply, marked as answer]
If you're using Postgresql, you can try deleting the Enum (called Type in PG) in the PGAdmin. See my screenshot in previous post. That's the only thing that worked for me.

**Reply by AhmadNadeemJB** [reply]
> If you're using Postgresql, you can try deleting the Enum (called Type in PG) in the PGAdmin. See my screenshot in previous post. That's the only thing that worked for me.

Thanks, worked for me.
